docs/design/README.md

```diff
@@ -0,0 +1,177 @@
+
+## Dependency handling
+
+In compiled programming languages, there are two kinds of dependencies: static and dynamic. Static are included statically and compiled when compiling the project,
+dynamic are pre-compiled libraries linked to the project.
+Note that Nushell is [similar to compiled languages][Nushell compiled] rather than typical dynamic languages like Python, so these concepts are relevan for Nushell.
+
+Static dependencies:
+* Advantages: reproducible, does not rely on system files (no more missing random.so.2), higher performance (allows joint optimization of dependencies and project itself)
+* Disadvatage: increased compile time, binary size, can easily end up with multiple versions of the same library (hello Nushell dependencies)
+
+Dynamic dependencies are the opposite basically. Note that Nushell currently supports only static dependencies, but we might be able to add the "linking" feature at some point.
+
+We might want `nupm`support both types of dependencies.
+
+## Package repository 
+
+Packages need to be stored somewhere. There should be one central "official" location (see https://github.com/NixOS/nixpkgs for inspiration).
+
+Additionally, user should be able to add 3rd party repositories as well as install local and other packages (e.g., from the web, just pointing at URL),
+as long as it has `METADATA_FILE` telling `nupm` what to do.
+
+## API / CLI Interface
+
+Nushell's module design conflates CLI interface with API -- they are the same. Not all of the below are of the same priority.
+
+> **Note**  
+> commands like `list`, `install`, `search`, `uninstall`, `update`, ..., i.e. should
+> - give short descriptions by default
+> - give long descriptions with `--long-description (-l)`
+
+- `nupm new [--script] [--module]`
+    - create a new local package with template files ([`kubouch/nuun`])
+- `nupm list`
+    - list currently installed packages and if they're out of date
+- `nupm install`
+    - install package into the currently active overlay (can override which overlay to install to)
+    - `--reinstall (-r)`: reinstall package if installed
+    - `--update (-u)`: update local packages if outdated
+    - `-u` > `-r`
+    - `--yes (-y)`: do not ask for user confirmation to use `nupm install` in scripts
+- `nupm add`
+    - add a dependency to the current project
+    - it is different from `nupm install`: this one adds the dependency to the MANIFEST_FILE, `nupm install` does not
+- `nupm uninstall`
+    - uninstall a package from a currently active overlay (can override which overlay to install to)
+    - `--yes (-y)`: do not ask for user confirmation to use `nupm uninstall` in scripts
+- `nupm update`
+    - update all packages in a currently active overlay (can specify package and/or overlay name)
+    - can be used to self-update: `nupm update nupm`, `nupm update --self` or `nupm update --all` (would update everything)
+    - `--yes (-y)`: do not ask for user confirmation to use `nupm update` in scripts
```

Comment:
```suggestion
    - `--yes (-y)`: do not ask for user confirmation
- `nupm update`
    - update all packages in a currently active overlay (can specify package and/or overlay name)
    - can be used to self-update: `nupm update nupm`, `nupm update --self` or `nupm update --all` (would update everything)
    - `--yes (-y)`: do not ask for user confirmation
```

The same case as above

Reply:
done in done in ad6b0cd8a06722ce05f90a2f67bf782f40c87010

i've just made the "script" an example